src/v/kafka/server/quota_manager.cc

```diff
@@ -158,12 +178,37 @@ throttle_delay quota_manager::record_tp_and_throttle(
 
     auto prev = it->second.delay;
     it->second.delay = delay_ms;
-
     throttle_delay res{};
     res.first_violation = prev.count() == 0;
     res.duration = it->second.delay;
     return res;
 }
+
+// record a new observation and return <previous delay, new delay>
+throttle_delay quota_manager::record_produce_tp_and_throttle(
+  std::optional<std::string_view> client_id,
+  uint64_t bytes,
+  clock::time_point now) {
+    auto it = maybe_add_and_retrieve_quota(client_id, now);
+
+    it->second.tp_produce_rate.record(bytes, now);
+
+    return throttle(client_id, now, request_type::produce);
```

Comment:
(discuss) `maybe_add_and_retrieve_quota()` is called twice while `record_produce_tp_and_throttle()` is executed: from here and from `throttle()`. Is there a way to have it called only once?

Reply:
fixed